### DISCUSSION TITLE
Is there a way to force path-style requests for AWS/MinIO?

### DISCUSSION THREAD

**Discussion by snowpong** [ropensci/targets#1134] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I'm currently switching to using a local MinIO server as my "aws" repository. Targets with format="file" are properly uploaded to MinIO, but they fail when downloading with "Could not resolve host: mybucket.s3.NULL.amazonaws.com". 

So essentially there are two ways to do S3 requests, it's either the path-style or virtual-host-style. The url above is virtual-host-style. Now, MINIO by default only support path-style out of the box. So, I can either enable virtual-host-style by setting [MINIO_DOMAIN](https://min.io/docs/minio/linux/reference/minio-server/minio-server.html#envvar.MINIO_DOMAIN) or I can make paws force path-style in the config with [s3_force_path_style](https://www.paws-r-sdk.com/docs/s3/#arguments).

However, when trying to force path-style with `s3_force_path_style = TRUE` in tar_resources_aws(..) I'm not allowed. It does a check and says it's an invalid option. Any ideas here?

Also, why on earth it's using amazonaws.com is beyond me. I have set the endpoint to "http://127.0.0.1:9000" and it's using that for uploads.

**Comment by wlandau** [comment, marked as answer]
As of a91ec548ceb228618b07fb2a67f4e9930abf17d8, I added a `s3_force_path_style` argument to `tar_resources_aws()`. Please let me know if that does not work for you.

**Reply by snowpong** [reply]
Amazing @wlandau I will let you know how it works!

**Reply by snowpong** [reply]
Thank you! I successfully managed to set s3_force_path_style and it changes the url it uses from virtual-host-style to path-style. So the change seems to do what I wanted.

However, it's still failing with:  `could not load dependencies of target some_file_user. Could not resolve host: s3.NULL.amazonaws.com` Here is the full output:

```
Rscript -e "targets::tar_make()"
✔ skip target some_file
Could not resolve host: s3.NULL.amazonaws.com
✔ skip target data
▶ start target some_file_user
✖ error target some_file_user
▶ end pipeline [1.949 seconds]
Warning message:
Please supply an explicit prefix for you target object data and metadata. The prefix should be unique to your {targets} project. In the future, {targets} will begin requiring explicitly user-supplied prefixes. This warning was added on 2023-08-24 ({targets} version 1.2.2.9000). 
Error:
! Error running targets::tar_make()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: could not load dependencies of target some_file_user. Could not resolve host: s3.NULL.amazonaws.com
Execution halted
make: *** [Makefile:2: all] Error 1
```

I don't understand where it gets this S3 endpoint from - the aws_resources I use in my tar_options is:

```
<tar_resources_aws>
  bucket: minio-targets-experiment
  prefix: _targets
  region: NULL
  part_size: 5242880
  endpoint: http://127.0.0.1:9000
  max_tries: NULL
  seconds_timeout: NULL
  close_connection: NULL
  s3_force_path_style: TRUE
  args: list()
> 
```

My guess is that 1) either the target itself has this S3 endpoint encoded in meta for some reason, or 2) paws is ignoring the endpoint provided, or 3) paws is not provided my endpoint for this particular call.

I will investigate further.

**Reply by wlandau** [reply]
Maybe you could try also setting the region in tar_resources_aws().

**Reply by wlandau** [reply]
`region = "auto"` may work.
